File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        res = 0 
        n = len(heights)
        # find left max 
        # find right max 
        # Compute max iteratively 

        i,j = 0, n-1
        while i < j:
            area = (j-i)*min(heights[i], heights[j])
            res = max(area, res)
            if heights[i] < heights[j]:
                i += 1
            else:
                j -= 1
        return res

        
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.go
func productExceptSelf(nums []int) []int {

	n := len(nums)
	left := make([]int, n)
	right := make([]int, n)
	res := make([]int, n)

	left[0] = nums[0] 
	right[n-1] = nums[n-1]

	for i := 1; i < n; i++ {
		left[i] = left[i-1] * nums[i]
	}
	for i := n-2; i >=0; i-- {
		right[i] = right[i+1] * nums[i]
	}

	for i := 0; i < n; i++ {

		if i == 0 {
			res[i] = right[i+1]
		} else if i == n-1 {
			res[i] = left[i-1]
		} else {
			res[i] = left[i-1]*right[i+1]
		}

		
	}

	return res
}
